intake_thredds/source.py

```diff
@@ -0,0 +1,70 @@
+import fnmatch
+
+from intake_xarray.base import DataSourceMixin
+
+from .cat import ThreddsCatalog
+
+try:
+    from tqdm import tqdm
+except ImportError:
+    tqdm = None
+
+
+class THREDDSMergedSource(DataSourceMixin):
+    version = '1.0'
+    container = 'xarray'
+    name = 'thredds_merged'
+    partition_access = True
+
+    def __init__(self, url, path, progressbar=True, metadata=None):
+        """
+
+        Parameters
+        ----------
+        url : str
+            Location of server
+        path : list of str
+            Subcats to follow; include glob characters (*, ?) in here for
+            matching
+        progressbar : bool
+            If True, will print a progress bar. Requires `tqdm <https://github.com/tqdm/tqdm>`__
+            to be installed.
+        metadata : dict or None
+            To associate with this source
+        """
+        super(THREDDSMergedSource, self).__init__(metadata=metadata)
+        self.urlpath = url
+        self.path = path
+        self._ds = None
+        self.progressbar = progressbar
+        if self.progressbar and tqdm is None:
+            raise ValueError("Missing package 'tqdm' required for progress bars.")
+
+    def _open_dataset(self):
+        import xarray as xr
+
+        if self._ds is None:
+            cat = ThreddsCatalog(self.urlpath)
+            for i in range(len(self.path)):
+                part = self.path[i]
+                if '*' not in part and '?' not in part:
+                    cat = cat[part]()
+                else:
+                    break
+            path = self.path[i:]
+            if self.progressbar:
+                data = [ds.to_dask() for ds in tqdm(_match(cat, path), desc='Dataset(s)', ncols=79)]
+            else:
+                data = [ds.to_dask() for ds in _match(cat, path)]
+            self._ds = xr.combine_by_coords(data)
+
+
+def _match(cat, patterns):
+    out = []
+    for name in cat:
+        if fnmatch.fnmatch(name, patterns[0]):
+            if len(patterns) == 1:
+                out.append(cat[name](chunks={}))
```

Comment:
Using `chunks={}` appears to speed the xarray's combine up:

```python
In [1]: import intake                                                                                              

In [2]: url = 'http://dap.nci.org.au/thredds/catalog.xml'                                                          

In [3]: paths = ['eMAST TERN', 'eMAST TERN - files', 'ASCAT', 'ASCAT_v1-0_soil-moisture_daily_0-05deg_2007-2011', '
   ...: 00000000', '*.nc']                                                                                         

In [4]: s = intake.open_thredds_merged(url, paths)                                                                 

In [5]: %%time 
   ...: ds = s.to_dask() 
   ...:  
   ...:                                                                                                            
Dataset(s): 100%|██████████████████████████████| 60/60 [02:36<00:00,  2.60s/it]
CPU times: user 875 ms, sys: 235 ms, total: 1.11 s
Wall time: 2min 41s
```


<img width="1011" alt="Screen Shot 2020-04-17 at 4 02 47 PM" src="https://user-images.githubusercontent.com/13301940/79617523-f0326000-80c4-11ea-9569-116a19cead92.png">